modules/dnn/src/op_cann.cpp

```diff
@@ -0,0 +1,322 @@
+{
+    CV_LOG_INFO(NULL, "Finalizing CANN Graph builder");
+    ge::aclgrphBuildFinalize();
+}
+
+void initAclGraphBuilder()
+{
+    using namespace ge;
+
+    static std::mutex mtx;
+    static bool aclGraphInitialized = false;
+    mtx.lock();
+    if (!aclGraphInitialized)
+    {
+        CV_LOG_INFO(NULL, "Initialize CANN Graph builder");
+        std::map<AscendString, AscendString> global_options = {
+            {AscendString(ir_option::SOC_VERSION), AscendString("Ascend310")},
+        }; // TODO: support other chips
+        ACL_CHECK_GRAPH_RET(aclgrphBuildInitialize(global_options));
+        aclGraphInitialized = true;
+    }
+    mtx.unlock();
+}
+
+bool CannGeGraph::loadFromONNX(const String& modelFile)
+{
+    using namespace ge;
+
+    std::map<AscendString, AscendString> parse_options; // TODO: support custom parse_options from file
+    CV_LOG_INFO(NULL, "Load graph from " << modelFile << " with ge::aclgrphParseONNX");
+    ACL_CHECK_GRAPH_RET(aclgrphParseONNX(modelFile.c_str(), parse_options, graph));
+
+    return true;
+}
+
+bool CannGeGraph::loadFromONNXFromMem(const char* buffer, size_t length)
+{
+    using namespace ge;
+
+    std::map<AscendString, AscendString> parse_options; // // TODO: support custom parse_options from file
+    CV_LOG_INFO(NULL, "Load graph from buffer with ge::aclgrphParseONNXFromMem");
+    ACL_CHECK_GRAPH_RET(aclgrphParseONNXFromMem(buffer, length, parse_options, graph));
+
+    return true;
+}
+
+void CannClient::init(int deviceId)
+{
+    initAcl();
+
+    device_id = deviceId;
+
+    ACL_CHECK_RET(aclrtSetDevice(device_id));
+    ACL_CHECK_RET(aclrtCreateContext(&context, device_id));
+    ACL_CHECK_RET(aclrtCreateStream(&stream));
+
+    initAclGraphBuilder();
+}
+
+CannClient::~CannClient()
+{
+    CV_LOG_INFO(NULL, "In ~CannClient, inputs = " << inputs << ", outputs = " << outputs);
+    if (!model_desc)
+    {
+        CV_LOG_INFO(NULL, "[Failed] Tried to deconstruct CannClient but model is not loaded");
+        return;
+    }
+    // free datasets: inputs, outputs
+    if (inputs)
+    {
+        CV_LOG_INFO(NULL, "In ~CannClient: destroy inputs");
+        destroyDataset(&inputs);
+    }
+    if (outputs)
+    {
+        CV_LOG_INFO(NULL, "In ~CannClient: destroy outputs");
+        destroyDataset(&outputs);
+    }
+    // unload model
+    ACL_CHECK_RET(aclmdlUnload(model_id));
+    // destroy model_desc
+    ACL_CHECK_RET(aclmdlDestroyDesc(model_desc));
+    model_desc = nullptr;
+    CV_LOG_INFO(NULL, "[Success] Unloaded model (id=" << model_id << ")");
+
+    // destroy stream
+    if (stream != nullptr)
+    {
+        ACL_CHECK_RET(aclrtDestroyStream(stream));
+        stream = nullptr;
+    }
+    // destroy context
+    if (context != nullptr)
+    {
+        ACL_CHECK_RET(aclrtDestroyContext(context));
+        context = nullptr;
+    }
+    // reset device
+    if (context == nullptr && stream == nullptr)
+    {
+        ACL_CHECK_RET(aclrtResetDevice(device_id));
+    }
```

Comment:
There is a potential issue of reset the same device multiple times when we load multiple models into the deivce. Resetting a device multiple times will not trigger unexpecting exit but will leave error messages in the log of CANN runtime.

Possible solution: force using `device_id=0` and call `aclrtResetDevice` in `CannClient::finalize()`;